src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1020 @@
+	iter_param.ip_recx.rx_idx	= PARITY_INDICATOR |
+					  (entry->ae_cur_stripe.as_stripenum *
+						entry->ae_oca->u.ec.e_len);
+	iter_param.ip_recx.rx_nr	= entry->ae_oca->u.ec.e_len;
+
+	D_DEBUG(DB_TRACE, "Querying parity for stripe: %lu, offset: %lu\n",
+		entry->ae_cur_stripe.as_stripenum,
+		iter_param.ip_recx.rx_idx);
+
+	/* Query the parity */
+	rc = vos_iterate(&iter_param, VOS_ITER_RECX, false, &anchors,
+			 agg_recx_iter_pre_cb, NULL, entry, NULL);
+	/* entry->ae_par_extent.ape_epoch has been set to the parity extent's
+	 * epoch
+	 */
+	if (rc != 0)
+		goto out;
+
+	D_DEBUG(DB_TRACE, "Par query: epoch: %lu, offset: %lu, length: %lu\n",
+		entry->ae_par_extent.ape_epoch,
+		entry->ae_par_extent.ape_recx.rx_idx,
+		entry->ae_par_extent.ape_recx.rx_nr);
+
+	if (entry->ae_par_extent.ape_epoch > entry->ae_cur_stripe.as_hi_epoch &&
+	    entry->ae_par_extent.ape_epoch != ~(0ULL)) {
+		/* Parity newer than data and holes, so delete the replicas.
+		 */
+		update_vos = true;
+		write_parity = false;
+		goto out;
+	}
+
+	if ((entry->ae_par_extent.ape_epoch == ~(0ULL) &&
+	     agg_stripe_is_filled(entry, false)) ||
+	     agg_stripe_is_filled(entry, true)) {
+		/* Replicas constitute a full stripe. */
+		rc = agg_encode_local_parity(entry);
+		goto out;
+	}
+
+	if (entry->ae_par_extent.ape_epoch == ~(0ULL)) {
+		/* No parity, partial-stripe worth of replica, nothing to do */
+		update_vos = false;
+		goto out;
+	}
+out:
+	if (update_vos && rc == 0) {
+		entry->ae_cur_stripe.as_suffix_ext = agg_get_carry_under(entry);
+		rc = agg_update_vos(entry, write_parity);
+	}
+
+	agg_clear_extents(entry);
+	return rc;
+}
+
+/* returns the subrange of the RECX iterator's returned recx that lies within
+ * the current stripe.
+ */
+static daos_off_t
+agg_in_stripe(struct ec_agg_entry *entry, daos_recx_t *recx)
+{
+	unsigned int		len = entry->ae_oca->u.ec.e_len;
+	unsigned int		k = entry->ae_oca->u.ec.e_k;
+	daos_off_t		stripe = recx->rx_idx / (len * k);
+	daos_off_t		stripe_end = (stripe + 1) * len * k;
+
+	if (recx->rx_idx + recx->rx_nr > stripe_end)
+		return stripe_end - recx->rx_idx;
+	else
+		return recx->rx_nr;
+}
+
+/* Iterator call back sub-function for handling data extents.
+ */
+static int
+agg_data_extent(vos_iter_entry_t *entry, struct ec_agg_entry *agg_entry,
+		unsigned int *acts)
+{
+	struct ec_agg_extent	*extent = NULL;
+	daos_off_t		 cur_stripenum, this_stripenum;
+	int			 rc = 0;
+
+	this_stripenum = agg_stripenum(agg_entry, entry->ie_recx.rx_idx);
+	if (this_stripenum != agg_entry->ae_cur_stripe.as_stripenum) {
+		/* Iterator has reached next stripe */
+		if (agg_entry->ae_cur_stripe.as_extent_cnt) {
+			cur_stripenum = agg_entry->ae_cur_stripe.as_stripenum;
+			rc = agg_process_stripe(agg_entry);
+			if (rc)
+				D_ERROR("Process stripe returned "DF_RC"\n",
+					DP_RC(rc));
+		/* Error leaves data covered by replicas vulnerable to vos
+		 * delete, so don't advance coordination epoch.
+		 */
+			rc = 0;
```

Comment:
Because there is no code in place to manage epoch ranges between EC and VOS aggregation. The comment is there to ensure the error is not ignored when the coordination with VOS aggregation is implemented.

With no epoch coordination, replica extents are always vulnerable to errant deletion by VOS aggregation. So no way to handle this error until the next patch.